File: Patterns/Decorators.py
# Create a base concrete class for weapons
from abc import ABC, abstractmethod


# Create an abstract class for Tool interface
class Tool(ABC):
    @abstractmethod
    def damage(self) -> float:
        pass

    @abstractmethod
    def durability(self) -> float:
        pass

    @abstractmethod
    def speed(self) -> float:
        pass

    @abstractmethod
    def description(self) -> str:
        pass


# Concrete Tool Classes with Constants
class Sword(Tool):
    DAMAGE = 50.0
    DURABILITY = 25.0
    SPEED = 40.0

    def damage(self) -> float:
        return self.DAMAGE

    def durability(self) -> float:
        return self.DURABILITY

    def speed(self) -> float:
        return self.SPEED

    def description(self) -> str:
        return "Sword"


class Bow(Tool):
    DAMAGE = 30.0
    DURABILITY = 15.0
    SPEED = 10.0

    def damage(self) -> float:
        return self.DAMAGE

    def durability(self) -> float:
        return self.DURABILITY

    def speed(self) -> float:
        return self.SPEED

    def description(self) -> str:
        return "Bow"


class Hammer(Tool):
    DAMAGE = 40.0
    DURABILITY = 50.0
    SPEED = 10.0

    def damage(self) -> float:
        return self.DAMAGE

    def durability(self) -> float:
        return self.DURABILITY

    def speed(self) -> float:
        return self.SPEED

    def description(self) -> str:
        return "Hammer"


# Decorator Pattern: Able to add enchantments/upgrades
class ToolDecorator(Tool):
    def __init__(self, tool: Tool):
        self.tool = tool

    def damage(self) -> float:
        return self.tool.damage()

    def durability(self) -> float:
        return self.tool.durability()

    def speed(self) -> float:
        return self.tool.speed()

    def description(self) -> str:
        return self.tool.description()


# Define a decorator for Sharpness enchantment
class SharpnessDecorator(ToolDecorator):
    def __init__(self, tool: Tool, sharpness_value: float = 5.0):
        super().__init__(tool)
        self.sharpness_value = sharpness_value

    def damage(self) -> float:
        return (
            self.sharpness_value + self.tool.damage()
        )  # Adds extra damage for sharpness

    def durability(self) -> float:
        return self.tool.durability()  # No change to durability

    def speed(self) -> float:
        return self.tool.speed()  # No change to speed

    def description(self) -> str:
        return (
            self.tool.description()
            + f" with Sharpness (+{self.sharpness_value} damage)"
        )


# Define a decorator for Durability enchantment
class DurabilityDecorator(ToolDecorator):
    def __init__(self, tool: Tool, durability_value: float = 10.0):
        super().__init__(tool)
        self.durability_value = durability_value

    def damage(self) -> float:
        return self.tool.damage()  # No change to damage

    def durability(self) -> float:
        return self.durability_value + self.tool.durability()  # Adds extra durability

    def speed(self) -> float:
        return self.tool.speed()  # No change to speed

    def description(self) -> str:
        return self.tool.description() + f" with Durability (+{self.durability_value})"


# Define a decorator for Speed Boost enchantment
class SpeedBoostDecorator(ToolDecorator):
    def __init__(self, tool: Tool, speed_value: float = 10.0):
        super().__init__(tool)
        self.speed_value = speed_value

    def damage(self) -> float:
        return self.tool.damage()  # No change to damage

    def durability(self) -> float:
        return self.tool.durability()  # No change to durability

    def speed(self) -> float:
        return self.speed_value + self.tool.speed()  # Adds extra speed

    def description(self) -> str:
        return (
            self.tool.description() + f" with Speed Boost (+{self.speed_value} speed)"
        )


# Create a basic sword
basic_sword = Sword()
print("Basic Sword:")
print("Description:", basic_sword.description())
print("Damage:", basic_sword.damage())
print("Durability:", basic_sword.durability())
print("Speed:", basic_sword.speed())

# Add Sharpness and Durability to the sword with parameterized values
enchanted_sword = DurabilityDecorator(
    SharpnessDecorator(basic_sword, sharpness_value=7.0), durability_value=12.0
)
print("\nEnchanted Sword:")
print("Description:", enchanted_sword.description())
print("Damage:", enchanted_sword.damage())
print("Durability:", enchanted_sword.durability())
print("Speed:", enchanted_sword.speed())

# Add another layer of Speed Boost to the enchanted sword
fully_enchanted_sword = SpeedBoostDecorator(enchanted_sword, speed_value=15.0)
print("\nFully Enchanted Sword:")
print("Description:", fully_enchanted_sword.description())
print("Damage:", fully_enchanted_sword.damage())
print("Durability:", fully_enchanted_sword.durability())
print("Speed:", fully_enchanted_sword.speed())

File: Java/Patterns/Decorators.java
// Define the Tool interface
interface Tool {
    double damage();
    double durability();
    double speed();
    String description();
}

// Create concrete classes that are tools 
class Sword implements Tool {
    private static final double DAMAGE = 50.0;
    private static final double DURABILITY = 25.0;
    private static final double SPEED = 40.0;

    @Override
    public double damage() {
        return DAMAGE;
    }

    @Override
    public double durability() {
        return DURABILITY;
    }

    @Override
    public double speed() {
        return SPEED;
    }

    @Override
    public String description() {
        return "Sword";
    }
}

class Bow implements Tool {
    private static final double DAMAGE = 30.0;
    private static final double DURABILITY = 15.0;
    private static final double SPEED = 10.0;

    @Override
    public double damage() {
        return DAMAGE;
    }

    @Override
    public double durability() {
        return DURABILITY;
    }

    @Override
    public double speed() {
        return SPEED;
    }

    @Override
    public String description() {
        return "Bow";
    }
}

class Hammer implements Tool {
    private static final double DAMAGE = 40.0;
    private static final double DURABILITY = 50.0;
    private static final double SPEED = 10.0;

    @Override
    public double damage() {
        return DAMAGE;
    }

    @Override
    public double durability() {
        return DURABILITY;
    }

    @Override
    public double speed() {
        return SPEED;
    }

    @Override
    public String description() {
        return "Hammer";
    }
}

// Create an abstract class that accepts and is a tool
abstract class ToolDecorator implements Tool {
    protected Tool tool;

    public ToolDecorator(Tool tool) {
        this.tool = tool;
    }

    @Override
    public double damage() {
        return tool.damage();
    }

    @Override
    public double durability() {
        return tool.durability();
    }

    @Override
    public double speed() {
        return tool.speed();
    }

    @Override
    public String description() {
        return tool.description();
    }
}

// Create decorators to change the underlying behavior of tools
class SharpnessDecorator extends ToolDecorator {
    private final double sharpnessValue;

    public SharpnessDecorator(Tool tool, double sharpnessValue) {
        super(tool);
        this.sharpnessValue = sharpnessValue;
    }

    @Override
    public double damage() {
        return sharpnessValue + tool.damage(); // Adds extra damage
    }

    @Override
    public String description() {
        return tool.description() + " with Sharpness (+" + sharpnessValue + " damage)";
    }
}

class DurabilityDecorator extends ToolDecorator {
    private final double durabilityValue;

    public DurabilityDecorator(Tool tool, double durabilityValue) {
        super(tool);
        this.durabilityValue = durabilityValue;
    }

    @Override
    public double durability() {
        return durabilityValue + tool.durability(); // Adds extra durability
    }

    @Override
    public String description() {
        return tool.description() + " with Durability (+" + durabilityValue + ")";
    }
}

class SpeedBoostDecorator extends ToolDecorator {
    private final double speedValue;

    public SpeedBoostDecorator(Tool tool, double speedValue) {
        super(tool);
        this.speedValue = speedValue;
    }

    @Override
    public double speed() {
        return speedValue + tool.speed(); // Adds extra speed
    }

    @Override
    public String description() {
        return tool.description() + " with Speed Boost (+" + speedValue + " speed)";
    }
}

// Test the design patterns
public class Decorators {
    public static void main(String[] args) {
        // Test Sword
        Tool basicSword = new Sword();
        Tool enchantedSword = new DurabilityDecorator(new SharpnessDecorator(basicSword, 7.0), 12.0);
        Tool fullyEnchantedSword = new SpeedBoostDecorator(enchantedSword, 15.0);

        System.out.println("Fully Enchanted Sword:");
        System.out.println("Description: " + fullyEnchantedSword.description());
        System.out.println("Damage: " + fullyEnchantedSword.damage());
        System.out.println("Durability: " + fullyEnchantedSword.durability());
        System.out.println("Speed: " + fullyEnchantedSword.speed());

        // Test Bow
        Tool basicBow = new Bow();
        Tool enchantedBow = new SharpnessDecorator(basicBow, 5.0);
        Tool durableEnchantedBow = new DurabilityDecorator(enchantedBow, 10.0);

        System.out.println("\nDurable Enchanted Bow:");
        System.out.println("Description: " + durableEnchantedBow.description());
        System.out.println("Damage: " + durableEnchantedBow.damage());
        System.out.println("Durability: " + durableEnchantedBow.durability());
        System.out.println("Speed: " + durableEnchantedBow.speed());

        // Test Hammer
        Tool basicHammer = new Hammer();
        Tool speedyHammer = new SpeedBoostDecorator(basicHammer, 20.0);

        System.out.println("\nSpeedy Hammer:");
        System.out.println("Description: " + speedyHammer.description());
        System.out.println("Damage: " + speedyHammer.damage());
        System.out.println("Durability: " + speedyHammer.durability());
        System.out.println("Speed: " + speedyHammer.speed());
    }
}
